Tests Refactor - Phase 2

rodrigonull opened the pull request:
This PR depends on: https://github.com/RedHatInsights/insights-host-inventory/pull/655

Moving forward with the tests refactoring, I've updated all tests to start using pytest features like fixtures and parametrization.

The test coverage remains the same: 95%
The test suite execution was decreased by ~90%

Files refactored:
- [x] test_api_auth.py
- [x] test_api_delete.py
- [x] test_api_get.py
- [x] test_api_health.py
- [x] test_api_patch.py
- [x] test_api_post.py
- [x] test_api_put.py
- [x] test_db_model.py
- [x] test_host_dedup_logic.py
- [x] test_mq_service.py
- [x] test_payload_tracker.py
- [x] test_system_culling.py
- [x] test_tagging.py
- [x] test_validators.py
- [x] test_xjoin.py

The only missing file on the list is the **test_unit.py** which can be refactored in the following sprints.

Glutexo commented on tests/conftest.py:
```diff
@@ -0,0 +1,165 @@
+        auth_type="account_number",
+        skip_host_validation=False,
+    ):
+        with subtests.test(msg="Creating or updating hosts"):
```
I don’t think this is a subtest. These are intended for parametrization. [Reference](https://docs.python.org/3.6/library/unittest.html#distinguishing-test-iterations-using-subtests)

> When there are very small differences among your tests, for instance some parameters, unittest allows you to distinguish them inside the body of a test method using the subTest() context manager.

rodrigonull replied:
I've removed all asserts from the fixtures

Glutexo commented on tests/conftest.py:
```diff
@@ -0,0 +1,165 @@
+@pytest.fixture(scope="function")
+def get_host(flask_client, subtests):
+    def _get(url, status=200, **query_parameters):
+        with subtests.test(msg="Retrieving a host"):
```
Again, this is not a subtest.

rodrigonull replied:
I've removed all asserts from the fixtures

Glutexo commented on tests/conftest.py:
```diff
@@ -0,0 +1,165 @@
+
+@pytest.fixture(scope="function")
+def get_host(flask_client, subtests):
+    def _get(url, status=200, **query_parameters):
```
```suggestion
    def _get(url, expected_status=200, **query_parameters):
```

- Can we accept the parameters as a _dict_?

   ```suggestion
       def _get(url, expected_status=200, query_parameters):
   ```

- Should we use this function for all the various GET operations? We can rather create separate functions without the URL parameter or taking a collection of IDs.

Glutexo commented on tests/conftest.py:
```diff
@@ -0,0 +1,165 @@
+        with subtests.test(msg="Retrieving a host"):
+            url = inject_qs(url, **query_parameters)
+            response = flask_client.get(url, headers=get_required_headers())
+            assert response.status_code == status
```
I still don’t like the combination of retrieval and assertion in one fixture. We had it like that, but now we have chance to fix it. I’d rather explicitly assert in the tests.

Glutexo commented on tests/conftest.py:
```diff
@@ -0,0 +1,165 @@
+@pytest.fixture(scope="function")
+def get_host_from_db(flask_app):
+    def _get(host_id):
+        return Host.query.filter(Host.id == host_id).first()
```
```suggestion
        return Host.query.get(host_id)
```

rodrigonull commented on tests/conftest.py:
```diff
@@ -0,0 +1,165 @@
+                    test_url = inject_qs(base_url, **{paging_parameter: invalid_value})
+                    get_host(test_url, 400)
+
+    return _invalid_paging_parameters_test
```
Since I've removed a bunch of fixtures I think we don't need to move the HTTP fixtures to a new module for now.

rodrigonull commented on tests/conftest.py:
```diff
@@ -0,0 +1,165 @@
+
+@pytest.fixture(scope="function")
+def mock_env_token(monkeypatch):
+    monkeypatch.setenv("INVENTORY_SHARED_SECRET", SHARED_SECRET)
```
I agree.

I've moved some fixtures to a "normal" functions

rodrigonull commented on tests/conftest.py:
```diff
@@ -0,0 +1,165 @@
+
+@pytest.fixture(scope="function")
+def mock_env_token(monkeypatch):
+    monkeypatch.setenv("INVENTORY_SHARED_SECRET", SHARED_SECRET)
```
I've removed this fixture. Only a single test that was using it.

rodrigonull commented:
Hey @Glutexo, thanks a lot for your code review. I've just refactored the **test_api_post.py** again, this time I tried to make it simple, like we discussed last week. I've removed a bunch of fixtures that I've introduced in my previous refactor. So if you have a spare cycle I would like to hear your thoughts on this code change again.

Glutexo commented:
Interesting. On my machine, some tests are failing. Investigating.

Glutexo commented:
Some test are failing if the local timezone is not UTC.

rodrigonull commented on app/utils.py:
```diff
@@ -93,6 +93,14 @@ def external_id(self):
     def external_id(self, cf):
         self.__data["external_id"] = cf
 
+    @property
+    def system_profile(self):
+        return self.__data.get("system_profile", None)
+
+    @system_profile.setter
+    def system_profile(self, system_profile):
+        self.__data["system_profile"] = system_profile
+
```
@Glutexo We weren't using the HostWrapper class for the tests related to system profile, we were getting the system profile data directly from a dict.

Another question is: shouldn't we move this HostWrapper class to the test folder as we only use it in tests?

Glutexo replied:
Definitely. Or even better, kill the _HostWrapper_ class with fire. But that can be done as separate tiny steps. Not in this PR though. It has been lingering around for too long already.

rodrigonull commented on app/models.py:
```diff
@@ -101,7 +101,7 @@ def __init__(
         self._update_ansible_host(ansible_host)
         self.account = account
-        self.facts = facts
+        self.facts = facts or {}
```
I agree with you, we shouldn't change the executive code in this PR. It was necessary because some tests I changed to create the hosts using the DB model directly and if we don't set facts to the model it breaks the "_serialize_facts(facts)" function because it expects facts being a dict.

I can update the tests to create hosts using MQ instead of DB model, so this won't be a problem and I can revert this change on the executive code.

Glutexo replied:
I was thinking about just manually coalescing to `{}` in the tests when calling the _Host_ constructor. I went through the places where the constructor is called and there is no single place anything than `{}` is passed in. That means that the coalesce here is rather useless, but at the same time, it doesn’t do any harm. Let’s keep it.

In the long run, I’d like to get completetly rid of this constructor and we even have tickets for that. But that’s not anything that should be done now.

Glutexo commented:
Merged _master_ into this. There was a conflict caused by a new test from #684. Converted that test and refactored it a bit to fix style and make the code more readable.

Glutexo commented:
@rodrigonull I‘d like to ask you to tak a look at the newest patch. Mainly to check whether the new test and its fixtures/helpers fit into the new pytest paradigm. Please comment and/or fix if not and ping me or any of the code owners (@jharting, @shdunning) to merge this.

rodrigonull commented:
@Glutexo Looks good to me!

Glutexo commented:
Awesome! The fixtures are tricky beasts!

Glutexo commented:
Merging.
